### PR TITLE
Refactor cuDNN Convolution memory format and Conv-Bias-Relu code

### LINKED ISSUE #55301: Detached tensor in channels_last format fails in the next layer
## 🐛 Bug

When a channels_last format tensor is detached, it fails to operate on another model.

## To Reproduce

Here is an example code that produces an unexpected error.

```python
import torch
import torch.nn as nn
import torch.cuda.amp as amp

device = torch.device('cuda')
dtype = torch.float32
memory_format = torch.channels_last
# memory_format = torch.contiguous_format

model1 = nn.Conv2d(3,3,1,1).to(device=device, dtype=dtype, non_blocking=True, memory_format=memory_format)
model2 = nn.Conv2d(3,3,1,1).to(device=device, dtype=dtype, non_blocking=True, memory_format=memory_format)

input = torch.randn(1,3,4,4).to(device, dtype=dtype, memory_format=memory_format)

with amp.autocast():
    out1 = model1(input)
    out2 = model2(out1.detach())
```

Running the code produces the following error message.
```bash
Traceback (most recent call last):
  File "channels_last_detach.py", line 17, in <module>
    out2 = model2(out1.detach())
  File "/home/seungjun/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/seungjun/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 399, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/seungjun/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/conv.py", line 396, in _conv_forward
    self.padding, self.dilation, self.groups)
RuntimeError: set_sizes_and_strides is not allowed on a Tensor created from .data or .detach().
If your intent is to change the metadata of a Tensor (such as sizes / strides / storage / storage_offset)
without autograd tracking the change, remove the .data / .detach() call and wrap the change in a `with torch.no_grad():` block.
For example, change:
    x.data.set_(y)
to:
    with torch.no_grad():
        x.set_(y)
```

The above code runs fine when the memory_format is set to `torch.contiguous_format` but setting the memory format to `torch.channels_last` does not work.

## Expected behavior

The above code should successfully compute out2 without error.
When I ran the code with pytorch 1.7.1, it worked fine.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
```bash
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 460.67

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[pip3] torchaudio==0.8.0a0+e4e171a
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py37he8ac12f_0
[conda] mkl_fft                   1.2.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.19.2           py37h54aff64_0
[conda] numpy-base                1.19.2           py37hfa32c7d_0
[conda] pytorch                   1.8.1           py3.7_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchaudio                0.8.1                      py37    pytorch
[conda] torchvision               0.9.1                py37_cu111    pytorch
```

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version:  11.1.1 / 8.0
 - GPU models and configuration: RTX 2080 Ti
 - Any other relevant information:

## Additional context

In contrast to the error message, there is no resize/reshape etc. operation on the detached tensor.
I assume this could be a bug in the newly introduced restriction on the `detach` function.

cc @VitalyFedyunin @jamesr66a @ppwwyyxx

### PR DESCRIPTION
**Pull request by xwang233**
This PR makes several changes:

- Changed function `bool cudnn_conv_use_channels_last(...)` to `at::MemoryFormat cudnn_conv_suggest_memory_format(...)`
- Removed `resize_` in cudnn convolution code. Added a new overloading method `TensorDescriptor::set` that also passes the desired memory format of the tensor.
- Disabled the usage of double + channels_last on cuDNN Conv-Relu and Conv-Bias-Relu. Call `.contiguous(memory_format)` before passing data to cuDNN functions.
- Disabled the usage of cuDNN fused Conv-Bias-Relu in cuDNN < 8.0 version due to a CUDNN_STATUS_NOT_SUPPORTED error. Instead, use the native fallback path.
- Let Conv-Bias-Relu code respect the global `allow_tf32` flag.

From cuDNN document, double + NHWC is genenrally not supported.

Close https://github.com/pytorch/pytorch/pull/66968

Fix #55301

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck @jjsjann123 @mcarilli

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -840,8 +842,8 @@ at::Tensor _convolution(
 
   at::MemoryFormat backend_memory_format = at::MemoryFormat::Contiguous;
 
-  if (detail::getCUDAHooks().compiledWithCuDNN() && cudnn_conv_use_channels_last(input, weight)) {
-    backend_memory_format = (k == 5) ? at::MemoryFormat::ChannelsLast3d : at::MemoryFormat::ChannelsLast;
+  if (detail::getCUDAHooks().compiledWithCuDNN()) {
+    backend_memory_format = cudnn_conv_suggest_memory_format(input, weight);
```
Comment (eqy) [inline aten/src/ATen/native/Convolution.cpp:846]: Does the `k==5` condition not matter anymore? (e.g., is it covered by the ` can_use_cudnn_channels_last_3d` condition?)

Comment (xwang233) [reply aten/src/ATen/native/Convolution.cpp:846]: Ideally `k==5` should be a subset of `can_use_cudnn_channels_last_3d`. We can add another check to make it safer though

```diff
- if (can_use_cudnn_channels_last_3d)
+ if (can_use_cudnn_channels_last_3d && input.dim() == 5)
```

in the new `cudnn_conv_suggest_memory_format` function

**Comment by xwang233** [comment]
I've reverted the `resize_` changes. The issue in https://github.com/pytorch/pytorch/issues/55301 will unfortunately still be left open.

**Comment by ngimel** [comment]
> I've reverted the `resize_` changes. The issue in #55301 will unfortunately still be left open.

I think you can remove `resize_`, but you'd need to send `memory_format` as an argument to `raw_convolution*` and created descriptors sending memory_format as a second argument, then for the descriptors `fixSizeOneDimStride` will be called and they will be created correctly?

**Comment by ngimel** [comment]
Otherwise, this generally looks good, no `resize_` calls is a welcome change, and cudnn descriptors should still be correctly created due to `memory_format` arg. 
@jjsjann123 let me know if you have other concerns

**Comment by ngimel** [comment]
Hm, interesting that a different kernel is picked. Can you put in fallback for 7.6.5, always calling conv, and then relu?

**Comment by xwang233** [comment]
Hi @ngimel , this PR is ready for review.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**